### PR TITLE
[cli][start] Added manifest middleware

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

- Split out of https://github.com/expo/expo/pull/16517
- Pending https://github.com/expo/expo/pull/16557

# How

- Combines most logic from the legacy manifest middleware with the new manifest middleware.

Implements a hierarchy for middleware:
- `ExpoMiddleware`: common middleware used by all for ensuring the error response and instantiation is the same. This enforces that all of our middleware has the base requirement of corresponding to a path like '/manifest'.
  - `ManifestMiddleware`: Think of this like a middleware which dynamically generates the head element of an `index.html`. This reads the Expo config to generate the name, icon etc, creates the bundle URL location, and responds in a unified way that is shared across the subclasses.
    - `ClassicManifestMiddleware`: Technically still the main middleware, this uses legacy headers like `exponent-accept-signature`, and `Exponent-Server`. The main feature is the legacy signing technique which does not support offline mode, instead defaulting the user to anonymous.
    - `ExpoGoManifestHandlerMiddleware`: The newer middleware implementation (originally implemented by @wschurman), this accounts for the runtime version and other EAS properties. This middleware uses a newer system for signing the manifest that enables subsequent runs to be offline.
  - `RedirectRuntimeMiddleware`: Moved to https://github.com/expo/expo/pull/16559
  - `InterstitialPageMiddleware`: Moved to https://github.com/expo/expo/pull/16559

# Test Plan

- Unit tests.

### CODE REVIEW

File: packages/expo/cli/start/server/middleware/ManifestMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+      constructUrl: UrlCreator['constructUrl'];
+      isNativeWebpack?: boolean;
+    }
+  ) {
+    super(
+      projectRoot,
+      /**
+       * Only support `/`, `/manifest`, `/index.exp` for the manifest middleware.
+       * Returns true when the manifest middleware should handle the incoming server request.
+       * Exposed for testing.
```
Comment (EvanBacon) [inline packages/expo/cli/start/server/middleware/ManifestMiddleware.ts:63]:
```suggestion
```

File: packages/expo/cli/start/server/middleware/resolveAssets.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+      );
+      // Delete the field so Expo Go doesn't attempt to read it.
+      delete manifest.android.googleServicesFile;
+    }
+  }
+  if (manifest.ios?.googleServicesFile) {
+    try {
+      const contents = await fs.readFile(
+        path.resolve(projectRoot, manifest.ios.googleServicesFile),
+        'base64'
```
Comment (ide) [inline packages/expo/cli/start/server/middleware/resolveAssets.ts:46]: Is this correct for the iOS file to be encoded as ASCII Base64 and the Android file to be UTF-8?

File: packages/expo/cli/start/server/middleware/resolveAssets.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    assetSchemas.forEach((manifestField, index: number) =>
+      set(manifest, `${manifestField}Url`, urls[index])
+    );
+  } catch (e) {
+    if (e.localAssetPath) {
+      Log.warn(
+        `Unable to resolve asset "${e.localAssetPath}" from "${e.manifestField}" in your app.json or app.config.js`
+      );
+    } else {
+      Log.warn(`Warning: Unable to resolve manifest assets. Icons might not work. ${e.message}.`);
```
Comment (ide) [inline packages/expo/cli/start/server/middleware/resolveAssets.ts:121]: Fonts seem common too:
```suggestion
      Log.warn(`Warning: Unable to resolve manifest assets. Icons and fonts might not work. ${e.message}.`);
```

File: packages/expo/cli/start/server/middleware/ExpoGoManifestHandlerMiddleware.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        },
+        expoGo: expoGoConfig,
+        scopeKey,
+      },
+    };
+
+    const headers = this.getDefaultResponseHeaders();
+    if (requestOptions.acceptSignature && !shouldUseAnonymousManifest) {
+      // TODO: Should this also be memoized?
+      const manifestSignature = await signExpoGoManifestAsync(expoUpdatesManifest);
```
Comment (ide) [inline packages/expo/cli/start/server/middleware/ExpoGoManifestHandlerMiddleware.ts:95]: Memoizing it for the duration of the CLI process should be fine.

This said, we're working towards making it so signing requires an internet connection only every now and then in order to get a signing certificate that's valid for, say, a week. Once you have a valid signing certificate, you'll be able to sign development manifests while offline and be able to access scoped data in Expo Go without `@anonymous` or unsigned manifests.

**Review by ide — CHANGES_REQUESTED** [review]

File: packages/expo/cli/start/server/middleware/resolveAssets.ts (hunk 2 above)
Comment (EvanBacon) [reply packages/expo/cli/start/server/middleware/resolveAssets.ts:46]: https://github.com/expo/expo-cli/blob/bf8edaea7b9bda705ffe0bcc666203bc4879f96b/packages/xdl/src/ProjectAssets.ts#L41-L54

File: packages/expo/cli/start/server/middleware/ExpoMiddleware.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  constructor(protected projectRoot: string, protected supportedPaths: string[]) {}
+
+  /**
+   * Returns true when the middleware should handle the incoming server request.
+   * Exposed for testing.
+   */
+  _shouldContinue(req: ServerRequest): boolean {
+    return (
+      !!req.url &&
+      this.supportedPaths.includes(
```
[the remark this answers is not in the document]
Comment (EvanBacon) [reply packages/expo/cli/start/server/middleware/ExpoMiddleware.ts:20]: `supportedPaths` is of type array, this is equality matching.

File: packages/expo/cli/start/server/middleware/resolvePlatform.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * Extract the runtime platform from the server request.
+ * 1. Query param `platform`: `?platform=ios`
+ * 2. Header `expo-platform`: `'expo-platform': ios`
+ * 2. Legacy header `exponent-platform`: `'exponent-platform': ios`
+ *
+ * Returns first item in the case of an array.
+ */
+export function parsePlatformHeader(req: ServerRequest): string {
+  const url = req?.url ? parse(req.url, /* parseQueryString */ true) : null;
```
[the remark this answers is not in the document]
Comment (EvanBacon) [reply packages/expo/cli/start/server/middleware/resolvePlatform.ts:18]: `req` is not optional but `req.url` is optional.

[the remark this answers is not in the document]
Comment (ide) [reply packages/expo/cli/start/server/middleware/resolvePlatform.ts]: What's happening here is that Node uses `http.IncomingMessage` in two scenarios, when a server is listening and receives a request (the scenario here) and when Node itself acts as an HTTP client and gets a response back from another server. In the first scenario, `req.url` is always defined, but it's not defined in the second scenario.

In short, I think we could write `type ServerRequest = http.IncomingMessage & { url: string };` or use `req.url!`/`nullthrows(req.url)` at minimum.

File: packages/expo/cli/start/server/middleware/ClassicManifestMiddleware.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+};
+
+export class ClassicManifestMiddleware extends ManifestMiddleware {
+  public getParsedHeaders(req: ServerRequest): ParsedHeaders {
+    const platform = parsePlatformHeader(req) || 'ios';
+    assertRuntimePlatform(platform);
+    return {
+      platform,
+      acceptSignature: Boolean(req.headers?.['exponent-accept-signature']),
+      hostname: stripPort(req.headers?.['host']),
```
Comment (ide) [inline packages/expo/cli/start/server/middleware/ClassicManifestMiddleware.ts:31]: req.headers is always non-null here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e7f55d3765c3ad57a19f7e47cc8e8ca9faca9fc4/types/node/http.d.ts#L871

File: packages/expo/cli/start/server/middleware/resolvePlatform.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ * 1. Query param `platform`: `?platform=ios`
+ * 2. Header `expo-platform`: `'expo-platform': ios`
+ * 2. Legacy header `exponent-platform`: `'exponent-platform': ios`
+ *
+ * Returns first item in the case of an array.
+ */
+export function parsePlatformHeader(req: ServerRequest): string | null {
+  const url = req.url ? parse(req.url, /* parseQueryString */ true) : null;
+  const platform =
+    url?.query?.platform || req?.headers?.['expo-platform'] || req?.headers?.['exponent-platform'];
```
Comment (ide) [inline packages/expo/cli/start/server/middleware/resolvePlatform.ts:20]:
```suggestion
  const url = parse(req.url!, /* parseQueryString */ true);
  const platform =
    url.query?.platform || req.headers['expo-platform'] || req.headers['exponent-platform'];
```

**Review by ide — APPROVED** [review]

File: packages/expo/cli/start/server/middleware/ManifestMiddleware.ts [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+  /** Parse request headers into options. */
+  public abstract getParsedHeaders(req: ServerRequest): ParsedHeaders;
+
+  /** Store device IDs that were sent in the request headers. */
+  private async saveDevicesAsync(req: ServerRequest) {
+    const deviceIds = req.headers?.['expo-dev-client-id'];
+    if (deviceIds) {
+      await ProjectDevices.saveDevicesAsync(this.projectRoot, deviceIds).catch((e) =>
+        Log.error(e.stack)
```
Comment (byCedric) [inline packages/expo/cli/start/server/middleware/ManifestMiddleware.ts:130]: I think the error message/summary is always great to present to users, seems like that's being swallowed here right? Maybe it's better to add something [like this](https://github.com/expo/expo/pull/16559/files#diff-fb386925332777171063110d0764fb5db28ed815d9a3bd8dc9b3a55b3007b744R46) here too:
```js
Log.error(chalk.red(e.toString()) + (EXPO_DEBUG ? '\n' + chalk.gray(e.stack) : ''));
```

Or even something like `Log.exception` which does exactly that?

**Review by byCedric — APPROVED** [review]
Looking great!

**Pull request merged**